### ISSUE TITLE
Leverage React error boundaries to improve user experience of client side errors

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#943]
## Feature Description

Currently if an error is raised within React and not caught, it causes the entire component tree to be unmounted, resulting in a blank screen which is a terrible experience for the user.

We should leverage React's [error boundaries](https://reactjs.org/docs/error-boundaries.html) at the top-level of our React apps to catch any errors that weren't caught, and then handle it more gracefully.

When catching an error, we could render a different component which could provide more information to the user, and better guide them as to what steps to take next. We could even give them information that they could copy and paste to provide when requesting support.

This gives the user a better experience and hopefully helps to provide better information for everyone to resolve the problem more quickly.

Error boundaries can be used at any level, (and we're already using them in some places) but this is particularly about using them as a last line of defense to prevent a blank screen.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Gracefully catch React errors on any of Site Kit's "React apps" (screens and admin bar).
* Show a message that an error occurred: _"Oops! Looks like this page is not rendering correctly. We'll help you get to the bottom of this — please start a thread on [our support forum](https://wordpress.org/support/plugin/google-site-kit/) and include the error snippet below."_
* Include a snippet of the error trace that people can copy and paste into e.g. the support forums.
* Use https://www.npmjs.com/package/@babel/plugin-transform-react-display-name to maintain component display names so that the error trace is more helpful in figuring out and resolving the issue.

## Implementation Brief

* Create a new `ErrorHandler` component that incorporates React Error Boundaries and the text above
* Wrap each of the "App" components in the `ErrorHandler` component:
  * assets/js/components/dashboard-details/dashboard-details-app.js
  * assets/js/components/dashboard-splash/dashboard-splash-app.js
  * assets/js/components/dashboard/dashboard-app.js
  * assets/js/components/settings/settings-app.js
* Add `@babel/plugin-transform-react-display-name` and ensure that component names are preserved in production builds
* Add a "testing"/debug error component to simulate the error. 😄 

## Changelog entry

* Improve JS error handling consistently across individual React apps and allow for better contextualization.

## QA Instructions

To test this import the test component that throws an error:

```js
import ErrorComponent from 'GoogleComponents/ErrorHandler/ErrorComponent';
```

and insert `<ErrorComponent />` inside any component other than `ErrorHandler` itself. It should display an error like:

<img width="1156" alt="Screenshot 2020-02-12 at 19 17 28" src="https://user-images.githubusercontent.com/90871/74368856-51d80480-4dcc-11ea-9ede-3821ce32f794.png">

**Comment by tofumatt** [comment]
All this sounds good. I'm assuming the filters shouldn't mess with the error boundaries at all.

Creating a simple "Oops, we encountered an error" screen that includes the full error a user could copy/paste would be handy as well.

We may want to see about getting our components to have preserved `displayName` properties to improve debugging (using https://www.npmjs.com/package/@babel/plugin-transform-react-display-name). This will make the errors we encounter more useful in general, and especially in this component.

**Comment by felixarntz** [comment]
IB ✅

**Comment by felixarntz** [comment]
@tofumatt Can you add an estimate? :)

**Comment by tofumatt** [comment]
Whoops! Done!

**Comment by tofumatt** [comment]
Investigating this further, it seems we use `componentDidCatch` in many of our components already: https://github.com/google/site-kit-wp/blob/0dbab5392a4234debf0a9d4e50bf3bd2fb002622/assets/js/googlesitekit-wp-dashboard.js#L53

What's weird is that disabling those checks still catches errors like uncaught `throw new Error( 'foo' );` or `const badArray = null; return badArray[1];`. I get console output like:

<img width="1395" alt="Screenshot 2019-12-10 at 20 21 38" src="https://user-images.githubusercontent.com/90871/70565765-acfe3880-1b8a-11ea-821e-6da48dfc2fb3.png">

and the page still works. I'm digging into this further, but I'm wondering if our filters or something have a giant `try/catch` block.

**Comment by tofumatt** [comment]
Confusingly, removing _all_ error handling is still causing errors to be caught. I'm curious if this is related to `doAction()`, because we aren't even using `console.error` anywhere else...

**Comment by felixarntz** [comment]
Removing this from 1.2.0 milestone, since I feel it's a bit too late for this change now.

**Comment by tofumatt** [comment]
Agreed. The existing error handling is a bit off and some unexpected behaviour is happening here. It'll require a bit more investigation.

**Comment by cole10up** [comment]
Quick smoke test of the app after building with this enhancement functioned. Looking for engineering to validate the error and send it to approval. @tofumatt @aaemnnosttv

**Comment by aaemnnosttv** [comment]
## QA ⚠️ 

* ✅ Added `<ErrorComponent>`  
![image](https://user-images.githubusercontent.com/1621608/75431357-6df3ae00-5955-11ea-9a12-ba5f2b7484f5.png)
* ✅ Clicking "Simulate an error" triggers error boundary displaying message  
![image](https://user-images.githubusercontent.com/1621608/75431456-8cf24000-5955-11ea-87d3-5f90ffc3f160.png)
* ⚠️ Console shows error about prop type passed to `Notification`  
    ```
    Warning: Failed prop type: Invalid prop `description` of type `object` supplied to `Notification`, expected `string`.
    ```  
    ![image](https://user-images.githubusercontent.com/1621608/75431273-4ac8fe80-5955-11ea-831d-fc698c8586b3.png)

@tofumatt - this will cause problems with the use of added `dompurify` on the notification description as well (passing JSX rather than string). We should probably either leave that as a plain message, or pass it as a string and add `code` to the element whitelist. What do you think?

**Comment by tofumatt** [comment]
Oh, smart. Yeah, I think having the error appear as `<code>` is helpful here. Why not update the description's `propType` here instead though? `PropTypes.element` should accept both a string an a JSX component, so we can pass strings or the `<code>{message}</code>` we do in https://github.com/google/site-kit-wp/pull/976/files#diff-954126ff8fcf4c3b8e5f715256bbb94aR46

Functionally JSX will happily render the description as either a string or element now, we just need to tweak propTypes.

I would rather check for the notification's description's type in the component and conditionally DOMPurify it (only run `sanitizeHTML` if it's a string) rather than passing it an HTML element. The former approach is _marginally_ more work, but much more flexible and won't box us in.

We should make that change to the Notification PR if it isn't merged; I can handle that super-quickly this morning if you like 😄

**Comment by aaemnnosttv** [comment]
Sounds great, let's do that 👍Thanks!

**Comment by tofumatt** [comment]
Done in https://github.com/google/site-kit-wp/pull/1170/commits/3812a03938cf2d1baa87045101e250c30fab60bc 👍

**Comment by tofumatt** [comment]
(I tested it myself and it worked with strings and React elements, also did the DOMPurification as expected with no PropType errors.)

**Comment by aaemnnosttv** [comment]
QA ✅ 

The only problem before was the propType error that was raised but did not affect the behavior of the error handler.

The `add/1110-remote-notifications` branch no longer raises PropType errors as well
![image](https://user-images.githubusercontent.com/1621608/75457659-68ae5780-5985-11ea-89e4-965d662c2884.png)

**Comment by aaemnnosttv** [comment]
QA ✅ 

The only problem before was the propType error that was raised but did not affect the behavior of the error handler.

This is no longer a problem after the fixes included in #1170 

![image](https://user-images.githubusercontent.com/1621608/75457659-68ae5780-5985-11ea-89e4-965d662c2884.png)

Moving to approval.
